src/Compilers/CSharp/Portable/Utilities/ValueSetFactory.DecimalTC.cs

```diff
@@ -4,13 +4,250 @@
 
 #nullable enable
 
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Threading;
+using Roslyn.Utilities;
+
 namespace Microsoft.CodeAnalysis.CSharp
 {
+    using static BinaryOperatorKind;
+
     internal static partial class ValueSetFactory
     {
-        private struct DecimalTC : EqualableValueTC<decimal>
+        private struct DecimalTC : INumericTC<decimal>
         {
-            decimal EqualableValueTC<decimal>.FromConstantValue(ConstantValue constantValue) => constantValue.DecimalValue;
+            // These are the smallest nonzero normal mantissa value (in three parts) below which you could use a higher scale.
+            // This is the 96-bit representation of ((2^96)-1) / 10;
+            private const uint transitionLow = 0x99999999;
+            private const uint transitionMid = 0x99999999;
+            private const uint transitionHigh = 0x19999999;
+            private const byte maxScale = 28;
+
+            private static readonly decimal normalZero = new decimal(lo: 0, mid: 0, hi: 0, isNegative: false, scale: maxScale);
+            private static readonly decimal epsilon = new decimal(lo: 1, mid: 0, hi: 0, isNegative: false, scale: maxScale);
```

Comment:
How did you choose epsilon?

Reply:
Smallest representable decimal.

---
In reply to: [383591184](https://github.com/dotnet/roslyn/pull/41912#discussion_r383591184) [](ancestors = 383591184)